applications: machine_learning: Implement power manager (#5672)
nrfconnect/sdk-nrf: doc/nrf/libraries/caf/buttons_pm_keep_alive.rst

.. _caf_buttons_pm_keep_alive:

> CAF: Buttons power manager keep alive module

Reviewer: ```suggestion
CAF: Power manager keep alive module for buttons
```
I guess this module is more related to power manager than to buttons? *Theoretically* speaking, there could be a similar module for LEDs?

> CAF: Power manager keep alive module for buttons

Reviewer 2: typo in commit message: `any configured button is pressed`

Author: Ok

> ############################################

Reviewer 2: Fix length of underscore made of `#`

Author: ok

.. contents::
   :local:
   :depth: 2

> The |buttons_pm_keep_alive| of the :ref:`lib_caf` (CAF) is responsible for generating :c:struct:`keep_alive_event` events on any configured button press.

Reviewer 2: `On any button press`? The module reacts on every button event.

> The |buttons_pm_keep_alive| of the :ref:`lib_caf` (CAF) is responsible for generating :c:struct:`keep_alive_event` events on any configured button press.

Author: But but it reacts only when pressed field is set:
```
        if (event->pressed) {
		keep_alive();
	}
```

> The |buttons_pm_keep_alive| of the :ref:`lib_caf` (CAF) is responsible for generating :c:struct:`keep_alive_event` events on any configured button press.

Author: @pdunaj - ok

> The |buttons_pm_keep_alive| of the :ref:`lib_caf` (CAF) is responsible for generating :c:struct:`keep_alive_event` event on any configured button press.

Reviewer: ```suggestion
The |buttons_pm_keep_alive| of the :ref:`lib_caf` (CAF) is responsible for generating :c:struct:`keep_alive_event` on any configured button press.
The :c:struct:`keep_alive_event` is used by the :ref:`CAF power manager module <caf_power_manager>` to reset the power-down counter when it is managing the overall operating system power state.
```
I think we need to mention the power manager here briefly -- after all, the module bears its name...

Configuration
*************

To use the module, you must enable the following Kconfig options:

> * :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` - This option enables |buttons_pm_keep_alive|.

Reviewer: ```suggestion
* :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` - This option enables the |buttons_pm_keep_alive|.
```

> * :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` - This option enables the |buttons_pm_keep_alive|.

Reviewer 2: Maybe we should inform that this option is selected by default when buttons and keep_alive_events are enabled ? So that user will be aware about it.

Author: Lets agree if we should enable it by default before any change in documentation.

Reviewer 2: Maybe we could state that this module is enabled by default if both buttons and power manager are enabled? No additional steps are needed to enable the module with current configuration.

Author: ok.

> * :kconfig:`CONFIG_CAF_BUTTONS` - This option enables the buttons module.

Reviewer 2: I wonder if we should depend on CAF buttons or on CAF button event here.
The module actually reacts also on button_events that can be provided by any module, CAF buttons is not necessary.

Author: Currently buttons_pm_keep_alive shares the configuration with buttons. Should I prepare another configuration that depends only buttons events instead of buttons module?

Reviewer 2: Other modules related to PM already share configuration with the main CAF module related to given functionality too. Let's leave it as it is for now.

> * :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` - This option enables the |buttons_pm_keep_alive| it is enabled by default if :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` and :kconfig:`CONFIG_CAF_BUTTONS` is enabled.

Reviewer 2: Something seems wrong here. btw. maybe we could simplify:

Set the :kconfig:`CONFIG_CAF_BUTTONS_PM_KEEP_ALIVE` Kconfig option to enable the |buttons_pm_keep_alive|.
The option is enabled by default if  :kconfig:`CONFIG_CAF_POWER_MANAGER` and :kconfig:`CONFIG_CAF_BUTTONS` are enabled.`

> * :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` - This option enables the |buttons_pm_keep_alive| it is enabled by default if :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` and :kconfig:`CONFIG_CAF_BUTTONS` is enabled.

Reviewer: ```suggestion
* :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` - This option enables the |buttons_pm_keep_alive|.
  It is enabled by default if :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` and :kconfig:`CONFIG_CAF_BUTTONS` are enabled.
```
Something wrong here -- :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE` is auto-referencing itself (enabled by default if it is itself enabled)...

Author: Right - it should be CAF_POWER_MANAGER

> * :kconfig:`CAF_KEEP_ALIVE_EVENTS` - This option enables keep alive event.

Reviewer 2: This is automatically enabled now, no need to enable it.

Author: Replaced by:
```
* :kconfig:`CAF_POWER_MANAGER` - This option enables power manager and is required by :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE`.
```

> * :kconfig:`CONFIG_CAF_BUTTONS` - This option enables the buttons module.

Reviewer: ```suggestion
* :kconfig:`CONFIG_CAF_BUTTONS` - This option enables the :ref:`CAF buttons module <caf_buttons>`.
```

> * :kconfig:`CAF_POWER_MANAGER` - This option enables power manager and is required by :kconfig:`CAF_BUTTONS_PM_KEEP_ALIVE`.

Reviewer: ```suggestion
* :kconfig:`CONFIG_CAF_POWER_MANAGER` - This option enables the :ref:`CAF power manager module <caf_power_manager>` and it is required by :kconfig:`CONFIG_CAF_BUTTONS_PM_KEEP_ALIVE`.
```

Implementation details
**********************

> The module reacts on :c:struct:`button_event` and generates :c:struct:`keep_alive_event` if any configured button is pressed.

Reviewer: Would it make sense to describe here briefly the relation of the module to the power manager module? Or maybe what can happen with the keep_alive_event once it is generated?

Author: I think it does not. User should look into keep_alive_event description for that. In other case we would have a kind of the documentation for keep_alive_event in different places in the repository. If anything updates or changes nobody would remember to change it here.

Reviewer: OK, and what about just mentioning the relation to power manager? After all, the module has power manager in its name and requires it to function. See my suggestion to the first line.

Author: It seems it was introduced by yours fix?

> The module reacts on :c:struct:`button_event` and generates :c:struct:`keep_alive_event` if any configured button is pressed.

Reviewer: ```suggestion
The module reacts to :c:struct:`button_event` and generates :c:struct:`keep_alive_event` for the CAF power manager if any configured button is pressed.
See :ref:`CAF buttons module <caf_buttons>` documentation page for information about how to configure buttons and the :ref:`CAF power manager module <caf_power_manager>` documentation page for more information about the power state management.
```
